kak-lsp.toml

```diff
@@ -159,15 +160,31 @@ args = [
     "--history-file=no",
     "-e",
     """
-        using LanguageServer;
-        using Pkg;
-        import StaticLint;
-        import SymbolServer;
-        import REPL;
-        env_path = dirname(Pkg.Types.Context().env.project_file);
-        server = LanguageServer.LanguageServerInstance(stdin, stdout, env_path, "");
-        server.runlinter = true;
-        run(server);
+    ls_install_path = joinpath(get(DEPOT_PATH, 1, joinpath(homedir(), ".julia")), "environments", "kak-lsp");
+    pushfirst!(LOAD_PATH, ls_install_path);
+    using LanguageServer;
+    popfirst!(LOAD_PATH);
+    depot_path = get(ENV, "JULIA_DEPOT_PATH", "");
+    buffer_file = ENV["kak_buffile"];
+    project_path = let 
+      dirname(something(
+        # 1. Check if there is an explicitly set project
+        Base.load_path_expand((
+          p = get(ENV, "JULIA_PROJECT", nothing);
+          p === nothing ? nothing : isempty(p) ? nothing : p
+        )),
+        # 2. Check for Project.toml in current working directory
+        Base.current_project(pwd()),
+        # 3. Check for Project.toml from buffer's full file path exluding the file name
+        Base.current_project(dirname(buffer_file)),
+        # 4. Fallback to global environment
+        Base.active_project()
```

Comment:
Why are there differences to the neovim config?

The "Configuration taken from https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/julials.lua"
comment is useful, we should keep it but use a permalink (commit SHA instead of "master").

Reply:
Opening a new *unsaved* buffer to file will cause the LSP to use a different environment. I experimented the default config from nvim-lspconfig, and created a file inside a directory with no `Project.toml`. What I noticed is that instead of falling back to default global environment, it assumed an **empty** project environment in `$HOME` instead of `$HOME/.julia/environments/<current-version-of-julia>`. So to fix the issue:
1. We have to check if there is a explicitly created `JULIA_PROJECT` environmental variable so when we do `JULIA_PROJECT=<path-to-project> kak newfile.jl`, it will use that environment regardless if the Project.toml is empty or not.
2. Use the current working directory as the basis of the current project environment.
3. If a file does not exist yet, we try using the path where we *will* create the file and use it as the basis for the project environment, hence, the `dirname` function from Julia's Base module.
4. We fallback to the global environment, hence, `Base.active_project()` *since we didn't use the `--project` flag in the arguments anyway*.

There are slight differences actually with the nvim-lspconfig but *I actually made my own lsp config instead of using theirs due to some experiences I have*. However, both follow the same goal through "hacky" workarounds. For example, these lines 

```julia
ls_install_path = joinpath(get(DEPOT_PATH, 1, joinpath(homedir(), ".julia")), "environments", "kak-lsp")
pushfirst!(LOAD_PATH, ls_install_path);
using LanguageServer;
popfirst!(LOAD_PATH);
```

It is a hacky workaround so we won't populate the LSP with symbols from the LanguageServer.jl and just use the plain environment of the global environment.

I hope this answers your questions. :smile: